File: back-end/src/main/java/es/ull/project/application/usecase/facility/CreateFacilityUseCase.java
package es.ull.project.application.usecase.facility;

import es.ull.project.domain.entity.Facility;
import es.ull.project.domain.enumerate.FacilityStatus;
import es.ull.project.domain.enumerate.FacilityType;
import es.ull.project.domain.valueobject.capacity.Capacity;
import es.ull.project.domain.valueobject.cost.OpeningFixedCost;
import es.ull.project.domain.valueobject.location.Location;

/**
 * Use case for creating a new facility.
 */
public interface CreateFacilityUseCase {
    /**
     * Creates a new facility with the specified attributes.
     *
     * @param facilityType      the type of facility
     * @param location          the location of the facility
     * @param capacity          the capacity of the facility
     * @param openingFixedCost  the fixed cost to open the facility
     * @param status            the initial status of the facility
     * @return the created facility
     */
    Facility create(FacilityType facilityType, Location location, Capacity capacity, OpeningFixedCost openingFixedCost, FacilityStatus status);
}

File: front-end/src/application/usecase/FacilityManagement/create-facility-use-case.ts
// CreateFacilityUseCase.ts
// Use case contract for creating a new facility

import type { Either, DataError } from '@ull-tfg/ull-tfg-typescript';
import { Facility } from '../../../domain/entity/facility';
import { FacilityType } from '../../../domain/enumerate/facility-type';
import { FacilityStatus } from '../../../domain/enumerate/facility-status';
import { Location } from '../../../domain/valueobject/location/location';
import { Capacity } from '../../../domain/valueobject/demand/capacity';
import { OpeningFixedCost } from '../../../domain/valueobject/cost/opening-fixed-cost';

/**
 * Use case for registering a new facility in the system.
 * Input: facilityType, location, capacity, openingFixedCost, status
 * Output: Facility (entity)
 */

// Command object for input data
export interface CreateFacilityCommand {
    facilityType: FacilityType;
    location: Location;
    capacity: Capacity;
    openingFixedCost: OpeningFixedCost;
    status: FacilityStatus;
}

// Result type for the use case
export type CreateFacilityResult = Facility;

// Use case contract
export interface CreateFacilityUseCase {
    /**
     * Handles the creation of a new facility
     *
     * @param command Data needed to create the facility
     * @returns Either a DataError or the created Facility
     */
    execute(command: CreateFacilityCommand): Promise<Either<DataError, CreateFacilityResult>>;
}
